In [ ]:
# 查看当前挂载的数据集目录
list.files ('/home/kesci/input/')

In [ ]:
# 查看个人持久化工作区文件
list.files ('/home/kesci/work/')

In [ ]:
# 查看当前kernerl下的package
pkgs <- installed.packages();
pkgs[is.na(pkgs[,"Priority"]), c("Package", "Version")]

In [ ]:
# 加载数据分析常用库
library(dplyr)
library(ggplot2)


# 数据类型

In [1]:
#数据类型
v <- TRUE 
print(class(v))#logic
v <- 23.5
print(class(v))#Num
v <- 2L
print(class(v))#int
v <- 2+5i
print(class(v))#complex
v <- "TRUE"
print(class(v))#Char
v <- charToRaw("Hello")
print(class(v))#Raw

[1] "logical"
[1] "numeric"
[1] "integer"
[1] "complex"
[1] "character"
[1] "raw"


In [2]:
# Create a vector.
apple <- c('red','green',"yellow")
print(apple)

[1] "red"    "green"  "yellow"


In [4]:
# Create a list.可以在其中包含许多不同类型的元素，如向量，函数甚至其中的另一个列表
list1 <- list(c(2,5,3),21.3,sin)
print(list1)

[[1]]
[1] 2 5 3

[[2]]
[1] 21.3

[[3]]
function (x)  .Primitive("sin")



In [5]:
# Create a matrix.
M = matrix( c('a','a','b','c','b','a'), nrow = 2, ncol = 3, byrow = TRUE)
print(M)

     [,1] [,2] [,3]
[1,] "a"  "a"  "b" 
[2,] "c"  "b"  "a" 


In [6]:
# Create an array.数组dim属性创建所需的维数
a <- array(c('green','yellow'),dim = c(3,3,2))
print(a)

, , 1

     [,1]     [,2]     [,3]    
[1,] "green"  "yellow" "green" 
[2,] "yellow" "green"  "yellow"
[3,] "green"  "yellow" "green" 

, , 2

     [,1]     [,2]     [,3]    
[1,] "yellow" "green"  "yellow"
[2,] "green"  "yellow" "green" 
[3,] "yellow" "green"  "yellow"



In [7]:
#Factors 因子
#使用向量创建的r对象。 它将向量与向量中元素的不同值一起存储为标签。 标签总是字符
# Create a vector.
apple_colors <- c('green','green','yellow','red','red','red','green')

# Create a factor object.
factor_apple <- factor(apple_colors)

# Print the factor.
print(factor_apple)
print(nlevels(factor_apple))

[1] green  green  yellow red    red    red    green 
Levels: green red yellow
[1] 3


In [8]:
#Data Frames 数据帧
#是表格数据对象。 与数据帧中的矩阵不同，每列可以包含不同的数据模式。 
#第一列可以是数字，而第二列可以是字符，第三列可以是逻辑的。
# Create the data frame.
BMI <- 	data.frame(
   gender = c("Male", "Male","Female"), 
   height = c(152, 171.5, 165), 
   weight = c(81,93, 78),
   Age = c(42,38,26)
)
print(BMI)

  gender height weight Age
1   Male  152.0     81  42
2   Male  171.5     93  38
3 Female  165.0     78  26


# 变量
有效的变量名称由字母，数字和点或下划线字符组成。 变量名以字母或不以数字后跟的点开头
可以使用print()或cat()函数打印变量的值。 cat()函数将多个项目组合成连续打印输出。

In [9]:
#查找变量
print(ls())


[1] "a"            "apple"        "apple_colors" "BMI"          "factor_apple"
[6] "list1"        "M"            "v"           


In [10]:
#匹配变量名
# List the variables starting with the pattern "var".
print(ls(pattern = "apple")) 

[1] "apple"        "apple_colors" "factor_apple"


In [11]:
#可以使用rm()函数删除变量。 下面我们删除变量var.3。 打印时，抛出变量错误的值。
rm(M)
print(M)


ERROR: Error in print(M): object 'M' not found


In [12]:
#清空变量
rm(list = ls())
print(ls())

character(0)


In [14]:
#------------------------运算符
#求余
v <- c( 2,5.5,6)
t <- c(8, 3, 4)
print(v%%t)

[1] 2.0 2.5 2.0


In [15]:
#将第二向量作为第一向量的指数	
v <- c( 2,5.5,6)
t <- c(8, 3, 4)
print(v^t)

[1]  256.000  166.375 1296.000


In [16]:
#---------赋值运算符
#这些运算符用于向向量赋值。
v1 <- c(3,1,TRUE,2+3i)
v2 <<- c(3,1,TRUE,2+3i)
v3 = c(3,1,TRUE,2+3i)
print(v1)
print(v2)
print(v3)
	
c(3,1,TRUE,2+3i) -> v1
c(3,1,TRUE,2+3i) ->> v2 
print(v1)
print(v2)

[1] 3+0i 1+0i 1+0i 2+3i
[1] 3+0i 1+0i 1+0i 2+3i
[1] 3+0i 1+0i 1+0i 2+3i
[1] 3+0i 1+0i 1+0i 2+3i
[1] 3+0i 1+0i 1+0i 2+3i


In [17]:
#此运算符用于将矩阵与其转置相乘。	
M = matrix( c(2,6,5,1,10,4), nrow = 2,ncol = 3,byrow = TRUE)
t = M %*% t(M)
print(t)

     [,1] [,2]
[1,]   65   82
[2,]   82  117


In [19]:
#----------包
#获取包含R包的库位置
.libPaths()

[1] "/opt/conda/lib/R/library"

In [20]:
library()

In [21]:
search()

[1] ".GlobalEnv"        "jupyter:irkernel"  "package:stats"    
 [4] "package:graphics"  "package:grDevices" "package:utils"    
 [7] "package:datasets"  "package:methods"   "Autoloads"        
[10] "package:base"

In [ ]:
# Install the package named "XML".
 install.packages("XML")

# 语句

In [1]:
# if语句
x <- 30L
if(is.integer(x)) {
   print("X is an Integer")
}

[1] "X is an Integer"


In [2]:
x <- c("what","is","truth")

if("Truth" %in% x) {
   print("Truth is found the first time")
} else if ("truth" %in% x) {
   print("truth is found the second time")
} else {
   print("No truth found")
}

[1] "truth is found the second time"


In [3]:
x <- switch(
   3,
   "first",
   "second",
   "third",
   "fourth"
)
print(x)

[1] "third"


In [4]:
# Repeat循环
v <- c("Hello","loop")
cnt <- 2

repeat {
   print(v)
   cnt <- cnt+1
   
   if(cnt > 5) {
      break
   }
}

[1] "Hello" "loop" 
[1] "Hello" "loop" 
[1] "Hello" "loop" 
[1] "Hello" "loop" 


In [5]:
#While循环
v <- c("Hello","while loop")
cnt <- 2

while (cnt < 7) {
   print(v)
   cnt = cnt + 1
}

[1] "Hello"      "while loop"
[1] "Hello"      "while loop"
[1] "Hello"      "while loop"
[1] "Hello"      "while loop"
[1] "Hello"      "while loop"


In [6]:
#For循环
v <- LETTERS[1:4]
for ( i in v) {
   print(i)
}

[1] "A"
[1] "B"
[1] "C"
[1] "D"


In [7]:
#next语句
v <- LETTERS[1:6]
for ( i in v) {
   
   if (i == "D") {
      next
   }
   print(i)
}

[1] "A"
[1] "B"
[1] "C"
[1] "E"
[1] "F"


# 数据处理

In [11]:
#---------------------------数据重塑
#rbind()函数合并两个数据帧。
# Create vector objects.
city <- c("Tampa","Seattle","Hartford","Denver")
state <- c("FL","WA","CT","CO")
zipcode <- c(33602,98104,06161,80294)
# Combine above three vectors into one data frame.
addresses <- cbind(city,state,zipcode)
# Print a header.
cat("# # #The First data frame") 
# Print the data frame.
print(addresses)

# # #The First data frame     city       state zipcode
[1,] "Tampa"    "FL"  "33602"
[2,] "Seattle"  "WA"  "98104"
[3,] "Hartford" "CT"  "6161" 
[4,] "Denver"   "CO"  "80294"


In [3]:
# Create another data frame with similar columns
#数据帧:x<-data.frame(a,b,c)
new.address <- data.frame(
   city = c("Lowry","Charlotte"),
   state = c("CO","FL"),
   zipcode = c("80230","33949"),
   stringsAsFactors = FALSE
)
# Print a header.
cat("# # # The Second data frame
") 
# Print the data frame.
print(new.address)
# Combine rows form both the data frames.
all.addresses <- rbind(addresses,new.address)

# Print a header.
cat("# # # The combined data frame") 

# Print the result.
print(all.addresses)

# # # The Second data frame
       city state zipcode
1     Lowry    CO   80230
2 Charlotte    FL   33949
# # # The combined data frame       city state zipcode
1     Tampa    FL   33602
2   Seattle    WA   98104
3  Hartford    CT    6161
4    Denver    CO   80294
5     Lowry    CO   80230
6 Charlotte    FL   33949


In [4]:
#merge()函数合并两个数据帧。 数据帧必须具有相同的列名称，在其上进行合并。
library(MASS)
merged.Pima <- merge(x = Pima.te, y = Pima.tr,
   by.x = c("bp", "bmi"),
   by.y = c("bp", "bmi")
)
print(merged.Pima)
nrow(merged.Pima)

   bp  bmi npreg.x glu.x skin.x ped.x age.x type.x npreg.y glu.y skin.y ped.y
1  60 33.8       1   117     23 0.466    27     No       2   125     20 0.088
2  64 29.7       2    75     24 0.370    33     No       2   100     23 0.368
3  64 31.2       5   189     33 0.583    29    Yes       3   158     13 0.295
4  64 33.2       4   117     27 0.230    24     No       1    96     27 0.289
5  66 38.1       3   115     39 0.150    28     No       1   114     36 0.289
6  68 38.5       2   100     25 0.324    26     No       7   129     49 0.439
7  70 27.4       1   116     28 0.204    21     No       0   124     20 0.254
8  70 33.1       4    91     32 0.446    22     No       9   123     44 0.374
9  70 35.4       9   124     33 0.282    34     No       6   134     23 0.542
10 72 25.6       1   157     21 0.123    24     No       4    99     17 0.294
11 72 37.7       5    95     33 0.370    27     No       6   103     32 0.324
12 74 25.9       9   134     33 0.460    81     No       8   126

[1] 17

In [5]:
#melt()拆分数据和cast()数据重构
library(MASS)
print(ships)

   type year period service incidents
1     A   60     60     127         0
2     A   60     75      63         0
3     A   65     60    1095         3
4     A   65     75    1095         4
5     A   70     60    1512         6
6     A   70     75    3353        18
7     A   75     60       0         0
8     A   75     75    2244        11
9     B   60     60   44882        39
10    B   60     75   17176        29
11    B   65     60   28609        58
12    B   65     75   20370        53
13    B   70     60    7064        12
14    B   70     75   13099        44
15    B   75     60       0         0
16    B   75     75    7117        18
17    C   60     60    1179         1
18    C   60     75     552         1
19    C   65     60     781         0
20    C   65     75     676         1
21    C   70     60     783         6
22    C   70     75    1948         2
23    C   75     60       0         0
24    C   75     75     274         1
25    D   60     60     251         0
26    D   60

In [9]:
install.packages("reshape")#安装数据包,否则无法用melt
library(reshape)
molten.ships <- melt(ships, id = c("type","year"))#将除类型和年份以外的所有列转换为多行展示。
print(molten.ships)

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


    type year  variable value
1      A   60    period    60
2      A   60    period    75
3      A   65    period    60
4      A   65    period    75
5      A   70    period    60
6      A   70    period    75
7      A   75    period    60
8      A   75    period    75
9      B   60    period    60
10     B   60    period    75
11     B   65    period    60
12     B   65    period    75
13     B   70    period    60
14     B   70    period    75
15     B   75    period    60
16     B   75    period    75
17     C   60    period    60
18     C   60    period    75
19     C   65    period    60
20     C   65    period    75
21     C   70    period    60
22     C   70    period    75
23     C   75    period    60
24     C   75    period    75
25     D   60    period    60
26     D   60    period    75
27     D   65    period    60
28     D   65    period    75
29     D   70    period    60
30     D   70    period    75
31     D   75    period    60
32     D   75    period    75
33     E  

In [10]:
#cast()重构数据
#使用cast()函数创建每年每种类型的船的总和
recasted.ship <- cast(molten.ships, type+year~variable,sum)
print(recasted.ship)

   type year period service incidents
1     A   60    135     190         0
2     A   65    135    2190         7
3     A   70    135    4865        24
4     A   75    135    2244        11
5     B   60    135   62058        68
6     B   65    135   48979       111
7     B   70    135   20163        56
8     B   75    135    7117        18
9     C   60    135    1731         2
10    C   65    135    1457         1
11    C   70    135    2731         8
12    C   75    135     274         1
13    D   60    135     356         0
14    D   65    135     480         0
15    D   70    135    1557        13
16    D   75    135    2051         4
17    E   60    135      45         0
18    E   65    135    1226        14
19    E   70    135    3318        17
20    E   75    135     542         1


# 函数

In [14]:

#function_name <- function(arg_1, arg_2, ...) {Function body }
# Create a function to print squares of numbers in sequence.
new.function <- function(a) {
   for(i in 1:a) {
      b <- i^2
      print(b)
   }
}
new.function(6)


[1] 1
[1] 4
[1] 9
[1] 16
[1] 25
[1] 36


In [15]:
# Create a function without an argument.
new.function <- function() {
   for(i in 1:5) {
      print(i^2)
   }
}	

# Call the function without supplying an argument.
new.function()

[1] 1
[1] 4
[1] 9
[1] 16
[1] 25


In [16]:
#使用参数值调用函数（按位置和名称）
# Create a function with arguments.
new.function <- function(a,b,c) {
   result <- a * b + c
   print(result)
}

# Call the function by position of arguments.
new.function(5,3,11)

# Call the function by names of the arguments.
new.function(a = 11, b = 5, c = 3)

[1] 26
[1] 58


# 字符串

In [19]:

#单双引号必须区分使用
a <- 'Start and end with single quote'
print(a)

b <- "Start and end with double quotes"
print(b)

c <- "single quote ' in between double quotes"
print(c)

d <- 'Double quotes " in between single quote'
print(d)

[1] "Start and end with single quote"
[1] "Start and end with double quotes"
[1] "single quote ' in between double quotes"
[1] "Double quotes \" in between single quote"


In [20]:
#连接字符串 - paste()函数,sep表示参数之间的任何分隔符。它是可选的。
#collapse用于消除两个字符串之间的空格。 但不是一个字符串的两个字内的空间。
a <- "Hello"
b <- 'How'
c <- "are you? "
print(paste(a,b,c))
print(paste(a,b,c, sep = "-"))
print(paste(a,b,c, sep = "", collapse = ""))

[1] "Hello How are you? "
[1] "Hello-How-are you? "
[1] "HelloHoware you? "


In [21]:
#格式化数字和字符串 - format()函数
#format(x, digits, nsmall, scientific, width, justify = c("left", "right", "centre", "none")) 
#nsmall是小数点右边的最小位数。
#科学设置为TRUE以显示科学记数法。
#width指示通过在开始处填充空白来显示的最小宽度。
#justify是字符串向左，右或中心的显示
# Total number of digits displayed. Last digit rounded off.
result <- format(23.123456789, digits = 9)
print(result)

# Display numbers in scientific notation.
result <- format(c(6, 13.14521), scientific = TRUE)
print(result)

# The minimum number of digits to the right of the decimal point.
result <- format(23.47, nsmall = 5)
print(result)

# Format treats everything as a string.
result <- format(6)
print(result)

# Numbers are padded with blank in the beginning for width.
result <- format(13.7, width = 6)
print(result)

# Left justify strings.
result <- format("Hello", width = 8, justify = "l")
print(result)

# Justfy string with center.
result <- format("Hello", width = 8, justify = "c")
print(result)

[1] "23.1234568"
[1] "6.000000e+00" "1.314521e+01"
[1] "23.47000"
[1] "6"
[1] "  13.7"
[1] "Hello   "
[1] " Hello  "


In [22]:
#计算字符串中的字符数 - nchar()函数
result <- nchar("Count the number of characters")
print(result)

[1] 30


In [23]:
#更改case - toupper()和tolower()函数
# Changing to Upper case.
result <- toupper("Changing To Upper")
print(result)

# Changing to lower case.
result <- tolower("Changing To Lower")
print(result)

[1] "CHANGING TO UPPER"
[1] "changing to lower"


In [24]:
#提取字符串的一部分 - substring()函数
#substring(x,first,last)
#x是字符向量输入。first是要提取的第一个字符的位置。last是要提取的最后一个字符的位置。
# Extract characters from 5th to 7th position.
result <- substring("Extract", 5, 7)
print(result)

[1] "act"


# 向量

In [25]:

# Creating a sequence from 5 to 13.
v <- 5:13
print(v)

[1]  5  6  7  8  9 10 11 12 13


In [26]:
# Create vector with elements from 5 to 9 incrementing by 0.4.
print(seq(5, 9, by = 0.4))

 [1] 5.0 5.4 5.8 6.2 6.6 7.0 7.4 7.8 8.2 8.6 9.0


In [27]:
# The logical and numeric values are converted to characters.
#使用C()函数
s <- c('apple','red',5,TRUE)
print(s)

[1] "apple" "red"   "5"     "TRUE" 


In [28]:
#访问向量元素
# Accessing vector elements using position.
t <- c("Sun","Mon","Tue","Wed","Thurs","Fri","Sat")
u <- t[c(2,3,6)]
print(u)

[1] "Mon" "Tue" "Fri"


In [29]:
# Accessing vector elements using logical indexing.
#使用索引访问向量的元素。 
#[]括号用于建立索引。 索引从位置1开始。
#在索引中给出负值会丢弃来自result.TRUE，FALSE或0和1的元素，也可用于索引。
v <- t[c(TRUE,FALSE,FALSE,FALSE,FALSE,TRUE,FALSE)]
print(v)

[1] "Sun" "Fri"


In [30]:
# Accessing vector elements using negative indexing.
x <- t[c(-2,-5)]
print(x)

[1] "Sun" "Tue" "Wed" "Fri" "Sat"


In [31]:
# Accessing vector elements using 0/1 indexing.
y <- t[c(0,0,0,0,0,0,1)]
print(y)

[1] "Sun"


In [32]:
# Create two vectors.
v1 <- c(3,8,4,5,0,11)
v2 <- c(4,11,0,8,1,2)

# Vector addition.
add.result <- v1+v2
print(add.result)

# Vector substraction.
sub.result <- v1-v2
print(sub.result)

# Vector multiplication.
multi.result <- v1*v2
print(multi.result)

# Vector division.
divi.result <- v1/v2
print(divi.result)

[1]  7 19  4 13  1 13
[1] -1 -3  4 -3 -1  9
[1] 12 88  0 40  0 22
[1] 0.7500000 0.7272727       Inf 0.6250000 0.0000000 5.5000000


In [33]:
#如果我们对不等长的两个向量应用算术运算，则较短向量的元素被循环以完成操作。
v1 <- c(3,8,4,5,0,11)
v2 <- c(4,11)
# V2 becomes c(4,11,4,11,4,11)

add.result <- v1+v2
print(add.result)

sub.result <- v1-v2
print(sub.result)

[1]  7 19  8 16  4 22
[1] -1 -3  0 -6 -4  0


In [34]:
#sort()函数排序。
v <- c(3,8,4,5,0,11, -9, 304)

# Sort the elements of the vector.
sort.result <- sort(v)
print(sort.result)

# Sort the elements in the reverse order.
revsort.result <- sort(v, decreasing = TRUE)
print(revsort.result)

# Sorting character vectors.
v <- c("Red","Blue","yellow","violet")
sort.result <- sort(v)
print(sort.result)

# Sorting character vectors in reverse order.
revsort.result <- sort(v, decreasing = TRUE)
print(revsort.result)

[1]  -9   0   3   4   5   8  11 304
[1] 304  11   8   5   4   3   0  -9
[1] "Blue"   "Red"    "violet" "yellow"
[1] "yellow" "violet" "Red"    "Blue"  


# 列表
列表是R语言对象，它包含不同类型的元素，如数字，字符串，向量和其中的另一个列表。 

In [35]:

# Create a list containing strings, numbers, vectors and a logical values.
list_data <- list("Red", "Green", c(21,32,11), TRUE, 51.23, 119.1)
print(list_data)

[[1]]
[1] "Red"

[[2]]
[1] "Green"

[[3]]
[1] 21 32 11

[[4]]
[1] TRUE

[[5]]
[1] 51.23

[[6]]
[1] 119.1



In [37]:
#列表元素可以给出名称，并且可以使用这些名称访问它们。
# Create a list containing a vector, a matrix and a list.
list_data <- list(c("Jan","Feb","Mar"), matrix(c(3,9,5,1,-2,8), nrow = 2),list("green",12.3))
# Give names to the elements in the list.
names(list_data) <- c("1st Quarter", "A_Matrix", "A Inner list")
# Show the list.
print(list_data)

$`1st Quarter`
[1] "Jan" "Feb" "Mar"

$A_Matrix
     [,1] [,2] [,3]
[1,]    3    5   -2
[2,]    9    1    8

$`A Inner list`
$`A Inner list`[[1]]
[1] "green"

$`A Inner list`[[2]]
[1] 12.3




In [39]:
#访问列表元素
# Access the thrid element. As it is also a list, all its elements will be printed.
print(list_data[3])
# Access the list element using the name of the element.
print(list_data$A_Matrix)

$`A Inner list`
$`A Inner list`[[1]]
[1] "green"

$`A Inner list`[[2]]
[1] 12.3


     [,1] [,2] [,3]
[1,]    3    5   -2
[2,]    9    1    8


In [40]:
#我们只能在列表的末尾添加和删除元素。 但我们可以更新任何元素。
# Create a list containing a vector, a matrix and a list.
list_data <- list(c("Jan","Feb","Mar"), matrix(c(3,9,5,1,-2,8), nrow = 2),
   list("green",12.3))

# Give names to the elements in the list.
names(list_data) <- c("1st Quarter", "A_Matrix", "A Inner list")

# Add element at the end of the list.
list_data[4] <- "New element"
print(list_data[4])

# Remove the last element.
list_data[4] <- NULL

# Print the 4th Element.
print(list_data[4])

# Update the 3rd Element.
list_data[3] <- "updated element"
print(list_data[3])

[[1]]
[1] "New element"

$<NA>
NULL

$`A Inner list`
[1] "updated element"



In [41]:
#合并列表
# Create two lists.
list1 <- list(1,2,3)
list2 <- list("Sun","Mon","Tue")

# Merge the two lists.
merged.list <- c(list1,list2)

# Print the merged list.
print(merged.list)

[[1]]
[1] 1

[[2]]
[1] 2

[[3]]
[1] 3

[[4]]
[1] "Sun"

[[5]]
[1] "Mon"

[[6]]
[1] "Tue"



In [42]:
#将列表转换为向量  使用unlist()函数
# Create lists.
list1 <- list(1:5)
print(list1)

list2 <-list(10:14)
print(list2)

# Convert the lists to vectors.
v1 <- unlist(list1)
v2 <- unlist(list2)

print(v1)
print(v2)

# Now add the vectors
result <- v1+v2
print(result)

[[1]]
[1] 1 2 3 4 5

[[1]]
[1] 10 11 12 13 14

[1] 1 2 3 4 5
[1] 10 11 12 13 14
[1] 11 13 15 17 19


# 矩阵
matrix(data, nrow行数, ncol列数, byrow, dimnames行列名称)



In [43]:
# Elements are arranged sequentially by row.
M <- matrix(c(3:14), nrow = 4, byrow = TRUE)
print(M)

# Elements are arranged sequentially by column.
N <- matrix(c(3:14), nrow = 4, byrow = FALSE)
print(N)

# Define the column and row names.
rownames = c("row1", "row2", "row3", "row4")
colnames = c("col1", "col2", "col3")

P <- matrix(c(3:14), nrow = 4, byrow = TRUE, dimnames = list(rownames, colnames))
print(P)


     [,1] [,2] [,3]
[1,]    3    4    5
[2,]    6    7    8
[3,]    9   10   11
[4,]   12   13   14
     [,1] [,2] [,3]
[1,]    3    7   11
[2,]    4    8   12
[3,]    5    9   13
[4,]    6   10   14
     col1 col2 col3
row1    3    4    5
row2    6    7    8
row3    9   10   11
row4   12   13   14


In [44]:
# Define the column and row names.
rownames = c("row1", "row2", "row3", "row4")
colnames = c("col1", "col2", "col3")

# Create the matrix.
P <- matrix(c(3:14), nrow = 4, byrow = TRUE, dimnames = list(rownames, colnames))

# Access the element at 3rd column and 1st row.
print(P[1,3])

# Access the element at 2nd column and 4th row.
print(P[4,2])

# Access only the  2nd row.
print(P[2,])

# Access only the 3rd column.
print(P[,3])

[1] 5
[1] 13
col1 col2 col3 
   6    7    8 
row1 row2 row3 row4 
   5    8   11   14 


In [45]:
# Create two 2x3 matrices.
matrix1 <- matrix(c(3, 9, -1, 4, 2, 6), nrow = 2)
print(matrix1)

matrix2 <- matrix(c(5, 2, 0, 9, 3, 4), nrow = 2)
print(matrix2)

# Add the matrices.
result <- matrix1 + matrix2
cat("Result of addition","
")
print(result)

# Subtract the matrices
result <- matrix1 - matrix2
cat("Result of subtraction","
")
print(result)

     [,1] [,2] [,3]
[1,]    3   -1    2
[2,]    9    4    6
     [,1] [,2] [,3]
[1,]    5    0    3
[2,]    2    9    4
Result of addition 
     [,1] [,2] [,3]
[1,]    8   -1    5
[2,]   11   13   10
Result of subtraction 
     [,1] [,2] [,3]
[1,]   -2   -1   -1
[2,]    7   -5    2


In [46]:
# Create two 2x3 matrices.
matrix1 <- matrix(c(3, 9, -1, 4, 2, 6), nrow = 2)
print(matrix1)

matrix2 <- matrix(c(5, 2, 0, 9, 3, 4), nrow = 2)
print(matrix2)

# Multiply the matrices.
result <- matrix1 * matrix2
cat("Result of multiplication","
")
print(result)

# Divide the matrices
result <- matrix1 / matrix2
cat("Result of division","
")
print(result)

     [,1] [,2] [,3]
[1,]    3   -1    2
[2,]    9    4    6
     [,1] [,2] [,3]
[1,]    5    0    3
[2,]    2    9    4
Result of multiplication 
     [,1] [,2] [,3]
[1,]   15    0    6
[2,]   18   36   24
Result of division 
     [,1]      [,2]      [,3]
[1,]  0.6      -Inf 0.6666667
[2,]  4.5 0.4444444 1.5000000


# 数组
数组是可以在两个以上维度中存储数据的R数据对象。

In [47]:
 
# Create two vectors of different lengths.
vector1 <- c(5,9,3)
vector2 <- c(10,11,12,13,14,15)

# Take these vectors as input to the array.
result <- array(c(vector1,vector2),dim = c(3,3,2))#3行3列两个
print(result)

, , 1

     [,1] [,2] [,3]
[1,]    5   10   13
[2,]    9   11   14
[3,]    3   12   15

, , 2

     [,1] [,2] [,3]
[1,]    5   10   13
[2,]    9   11   14
[3,]    3   12   15



In [48]:
#我们可以使用dimnames参数给数组中的行，列和矩阵命名。
# Create two vectors of different lengths.
vector1 <- c(5,9,3)
vector2 <- c(10,11,12,13,14,15)
column.names <- c("COL1","COL2","COL3")
row.names <- c("ROW1","ROW2","ROW3")
matrix.names <- c("Matrix1","Matrix2")

# Take these vectors as input to the array.
result <- array(c(vector1,vector2),dim = c(3,3,2),dimnames = list(row.names,column.names,
   matrix.names))
print(result)

, , Matrix1

     COL1 COL2 COL3
ROW1    5   10   13
ROW2    9   11   14
ROW3    3   12   15

, , Matrix2

     COL1 COL2 COL3
ROW1    5   10   13
ROW2    9   11   14
ROW3    3   12   15



In [49]:
#访问数组元素
# Print the third row of the second matrix of the array.
print(result[3,,2])

# Print the element in the 1st row and 3rd column of the 1st matrix.
print(result[1,3,1])

# Print the 2nd Matrix.
print(result[,,2])

COL1 COL2 COL3 
   3   12   15 
[1] 13
     COL1 COL2 COL3
ROW1    5   10   13
ROW2    9   11   14
ROW3    3   12   15


In [50]:
#操作数组元素
# Create two vectors of different lengths.
vector1 <- c(5,9,3)
vector2 <- c(10,11,12,13,14,15)

# Take these vectors as input to the array.
array1 <- array(c(vector1,vector2),dim = c(3,3,2))

# Create two vectors of different lengths.
vector3 <- c(9,1,0)
vector4 <- c(6,0,11,3,14,1,2,6,9)
array2 <- array(c(vector1,vector2),dim = c(3,3,2))

# create matrices from these arrays.
matrix1 <- array1[,,2]
matrix2 <- array2[,,2]

# Add the matrices.
result <- matrix1+matrix2
print(result)

     [,1] [,2] [,3]
[1,]   10   20   26
[2,]   18   22   28
[3,]    6   24   30


In [51]:
#跨数组元素的计算
#apply(数组, 数据集名称, 数组元素的函数)
# Create two vectors of different lengths.
vector1 <- c(5,9,3)
vector2 <- c(10,11,12,13,14,15)

# Take these vectors as input to the array.
new.array <- array(c(vector1,vector2),dim = c(3,3,2))
print(new.array)

# Use apply to calculate the sum of the rows across all the matrices.
result <- apply(new.array, c(1), sum)
print(result)


, , 1

     [,1] [,2] [,3]
[1,]    5   10   13
[2,]    9   11   14
[3,]    3   12   15

, , 2

     [,1] [,2] [,3]
[1,]    5   10   13
[2,]    9   11   14
[3,]    3   12   15

[1] 56 68 60


# 因子
因子是用于对数据进行分类并将其存储为级别的数据对象。 
它们可以存储字符串和整数。 它们在具有有限数量的唯一值的列中很有用。

In [52]:
# Create a vector as input.
data <- c("East","West","East","North","North","East","West","West","West","East","North")

print(data)
print(is.factor(data))

# Apply the factor function.
factor_data <- factor(data)

print(factor_data)
print(is.factor(factor_data))

 [1] "East"  "West"  "East"  "North" "North" "East"  "West"  "West"  "West" 
[10] "East"  "North"
[1] FALSE
 [1] East  West  East  North North East  West  West  West  East  North
Levels: East North West
[1] TRUE


In [53]:
# Create the vectors for data frame.
height <- c(132,151,162,139,166,147,122)
weight <- c(48,49,66,53,67,52,40)
gender <- c("male","male","female","female","male","female","male")

# Create the data frame.
input_data <- data.frame(height,weight,gender)
print(input_data)

# Test if the gender column is a factor.
print(is.factor(input_data$gender))

# Print the gender column so see the levels.
print(input_data$gender)

  height weight gender
1    132     48   male
2    151     49   male
3    162     66 female
4    139     53 female
5    166     67   male
6    147     52 female
7    122     40   male
[1] TRUE
[1] male   male   female female male   female male  
Levels: female male


In [54]:
#更改级别顺序
#可以通过使用新的等级次序再次应用因子函数来改变因子中的等级的顺序。
data <- c("East","West","East","North","North","East","West","West","West","East","North")
# Create the factors
factor_data <- factor(data)
print(factor_data)

# Apply the factor function with required order of the level.
new_order_data <- factor(factor_data,levels = c("East","West","North"))
print(new_order_data)


 [1] East  West  East  North North East  West  West  West  East  North
Levels: East North West
 [1] East  West  East  North North East  West  West  West  East  North
Levels: East West North


In [55]:
#生成因子级别
#gl(给出级数的整数, 复制数目的整数, 因子水平的标签向量)
v <- gl(3, 4, labels = c("Tampa", "Seattle","Boston"))
print(v)

 [1] Tampa   Tampa   Tampa   Tampa   Seattle Seattle Seattle Seattle Boston 
[10] Boston  Boston  Boston 
Levels: Tampa Seattle Boston


# 数据帧
数据帧是表或二维阵列状结构，其中每一列包含一个变量的值，
并且每一行包含来自每一列的一组值。

In [56]:

# Create the data frame.
emp.data <- data.frame(
   emp_id = c (1:5), 
   emp_name = c("Rick","Dan","Michelle","Ryan","Gary"),
   salary = c(623.3,515.2,611.0,729.0,843.25), 
   
   start_date = as.Date(c("2012-01-01", "2013-09-23", "2014-11-15", "2014-05-11",
      "2015-03-27")),
   stringsAsFactors = FALSE
)
# Print the data frame.			
print(emp.data) 

  emp_id emp_name salary start_date
1      1     Rick 623.30 2012-01-01
2      2      Dan 515.20 2013-09-23
3      3 Michelle 611.00 2014-11-15
4      4     Ryan 729.00 2014-05-11
5      5     Gary 843.25 2015-03-27


In [60]:
#通过使用str()函数可以看到数据帧的结构。
# Create the data frame.

# Get the structure of the data frame.
str(emp.data)


'data.frame':	5 obs. of  4 variables:
 $ emp_id    : int  1 2 3 4 5
 $ emp_name  : chr  "Rick" "Dan" "Michelle" "Ryan" ...
 $ salary    : num  623 515 611 729 843
 $ start_date: Date, format: "2012-01-01" "2013-09-23" ...


In [59]:
#用summary()函数获取数据的统计摘要和性质
# Create the data frame.

print(summary(emp.data))  

     emp_id    emp_name             salary        start_date        
 Min.   :1   Length:5           Min.   :515.2   Min.   :2012-01-01  
 1st Qu.:2   Class :character   1st Qu.:611.0   1st Qu.:2013-09-23  
 Median :3   Mode  :character   Median :623.3   Median :2014-05-11  
 Mean   :3                      Mean   :664.4   Mean   :2014-01-14  
 3rd Qu.:4                      3rd Qu.:729.0   3rd Qu.:2014-11-15  
 Max.   :5                      Max.   :843.2   Max.   :2015-03-27  


In [61]:
#提取特定列。
# Extract Specific columns.
result <- data.frame(emp.data$emp_name,emp.data$salary)
print(result)

  emp.data.emp_name emp.data.salary
1              Rick          623.30
2               Dan          515.20
3          Michelle          611.00
4              Ryan          729.00
5              Gary          843.25


In [62]:
#先提取前两行，然后提取所有列
# Extract first two rows.
result <- emp.data[1:2,]
print(result)

  emp_id emp_name salary start_date
1      1     Rick  623.3 2012-01-01
2      2      Dan  515.2 2013-09-23


In [63]:
# Extract 3rd and 5th row with 2nd and 4th column.
result <- emp.data[c(3,5),c(2,4)]
print(result)

  emp_name start_date
3 Michelle 2014-11-15
5     Gary 2015-03-27


In [64]:
#添加列
# Add the "dept" coulmn.
emp.data$dept <- c("IT","Operations","IT","HR","Finance")
v <- emp.data
print(v)

  emp_id emp_name salary start_date       dept
1      1     Rick 623.30 2012-01-01         IT
2      2      Dan 515.20 2013-09-23 Operations
3      3 Michelle 611.00 2014-11-15         IT
4      4     Ryan 729.00 2014-05-11         HR
5      5     Gary 843.25 2015-03-27    Finance


In [65]:
#添加行,并使用rbind()函数
# Create the second data frame
emp.newdata <- 	data.frame(
   emp_id = c (6:8), 
   emp_name = c("Rasmi","Pranab","Tusar"),
   salary = c(578.0,722.5,632.8), 
   start_date = as.Date(c("2013-05-21","2013-07-30","2014-06-17")),
   dept = c("IT","Operations","Fianance"),
   stringsAsFactors = FALSE
)

# Bind the two data frames.
emp.finaldata <- rbind(emp.data,emp.newdata)
print(emp.finaldata)

  emp_id emp_name salary start_date       dept
1      1     Rick 623.30 2012-01-01         IT
2      2      Dan 515.20 2013-09-23 Operations
3      3 Michelle 611.00 2014-11-15         IT
4      4     Ryan 729.00 2014-05-11         HR
5      5     Gary 843.25 2015-03-27    Finance
6      6    Rasmi 578.00 2013-05-21         IT
7      7   Pranab 722.50 2013-07-30 Operations
8      8    Tusar 632.80 2014-06-17   Fianance


# 条形图
barplot(H, xlab, ylab, main, names.arg, col)
* H是包含在条形图中使用的数值的向量或矩阵。
* xlab是x轴的标签。
* ylab是y轴的标签。
* main是条形图的标题。
* names.arg是在每个条下出现的名称的向量。
* col用于向图中的条形提供颜色。

In [25]:
# Create the data for the chart.
H <- c(7,12,28,3,41)
# Give the chart file a name.
#png(file = "barchart.png")
# Plot the bar chart.
barplot(H)
# Save the file.
dev.off()

png 
  3

plot without title

In [26]:
# Create the data for the chart.
H <- c(7,12,28,3,41)
M <- c("Mar","Apr","May","Jun","Jul")
# Give the chart file a name.
#png(file = "barchart_months_revenue.png")
# Plot the bar chart.
barplot(H,names.arg = M,xlab = "Month",ylab = "Revenue",col = "blue",
main = "Revenue chart",border = "red")
# Save the file.
dev.off()

png 
  3

Plot with title “Revenue chart”

In [27]:
#组合条形图和堆积条形图
# Create the input vectors.
colors <- c("green","orange","brown")
months <- c("Mar","Apr","May","Jun","Jul")
regions <- c("East","West","North")

# Create the matrix of the values.
Values <- matrix(c(2,9,3,11,9,4,8,7,3,12,5,2,8,10,11),nrow = 3,ncol = 5,byrow = TRUE)

# Give the chart file a name.
#png(file = "barchart_stacked.png")

# Create the bar chart.
barplot(Values,main = "total revenue",names.arg = months,xlab = "month",ylab = "revenue",
   col = colors)

# Add the legend to the chart.
legend("topleft", regions, cex = 1.3, fill = colors)

# Save the file.
dev.off()

png 
  3

Plot with title “total revenue”

# 箱线图
boxplot(x, data, notch, varwidth, names, main)
* x是向量或公式。

* data是数据帧。

* notch是逻辑值。 设置为TRUE以绘制凹口。

* varwidth是一个逻辑值。 设置为true以绘制与样本大小成比例的框的宽度。

* names是将打印在每个箱线图下的组标签。

* main用于给图表标题。

In [6]:
input <- mtcars[,c('mpg','cyl')]
print(head(input))

                   mpg cyl
Mazda RX4         21.0   6
Mazda RX4 Wag     21.0   6
Datsun 710        22.8   4
Hornet 4 Drive    21.4   6
Hornet Sportabout 18.7   8
Valiant           18.1   6


In [28]:
# Give the chart file a name.
#png(file = "boxplot.png")
# Plot the chart.
boxplot(mpg ~ cyl, data = mtcars, xlab = "Number of Cylinders",
   ylab = "Miles Per Gallon", main = "Mileage Data")
# Save the file.
dev.off()

png 
  3

Plot with title “Mileage Data”

In [29]:
#带槽的箱线图
# Give the chart file a name.
#png(file = "boxplot_with_notch.png")
# Plot the chart.
boxplot(mpg ~ cyl, data = mtcars, 
   xlab = "Number of Cylinders",
   ylab = "Miles Per Gallon", 
   main = "Mileage Data",
   notch = TRUE, 
   varwidth = TRUE, 
   col = c("green","yellow","purple"),
   names = c("High","Medium","Low")
)
# Save the file.
dev.off()

Warning message in bxp(list(stats = structure(c(21.4, 22.8, 26, 30.4, 33.9, 17.8, :
“some notches went outside hinges ('box'): maybe set notch=FALSE”

png 
  3

Plot with title “Mileage Data”

# 直方图
hist(v,main,xlab,xlim,ylim,breaks,col,border)
* v是包含直方图中使用的数值的向量。
* main表示图表的标题。
* col用于设置条的颜色。
* border用于设置每个条的边框颜色。
* xlab用于给出x轴的描述。
* xlim用于指定x轴上的值的范围。
* ylim用于指定y轴上的值的范围。
* break用于提及每个条的宽度。

In [30]:
# Create data for the graph.
v <-  c(9,13,21,8,36,22,12,41,31,33,19)

# Give the chart file a name.
#png(file = "histogram.png")

# Create the histogram.
hist(v,xlab = "Weight",col = "yellow",border = "blue")

# Save the file.
dev.off()


png 
  3

Plot with title “Histogram of v”

In [31]:
#X和Y值的范围
# Create data for the graph.
v <- c(9,13,21,8,36,22,12,41,31,33,19)

# Give the chart file a name.
#png(file = "histogram_lim_breaks.png")

# Create the histogram.
hist(v,xlab = "Weight",col = "green",border = "red", xlim = c(0,40), ylim = c(0,5),
   breaks = 5)

# Save the file.
dev.off()

png 
  3

Plot with title “Histogram of v”

# 折线图
plot(v,type,col,xlab,ylab)
* v是包含数值的向量。

* 类型采用值“p”仅绘制点，“l”仅绘制线和“o”绘制点和线。

* xlab是x轴的标签。

* ylab是y轴的标签。

* main是图表的标题。

* col用于给点和线的颜色。

In [32]:
# Create the data for the chart.
v <- c(7,12,28,3,41)

# Give the chart file a name.
#png(file = "line_chart.jpg")

# Plot the bar chart. 
plot(v,type = "o")

# Save the file.
dev.off()

png 
  3

plot without title

In [24]:
# Create the data for the chart.
v <- c(7,12,28,3,41)

# Give the chart file a name.
#png(file = "line_chart_label_colored.jpg")

# Plot the bar chart.
plot(v,type = "o", col = "red", xlab = "Month", ylab = "Rain fall",
   main = "Rain fall chart")

# Save the file.


dev.off()


png 
  3

Plot with title “Rain fall chart”

In [33]:
#多线型折线图
# Create the data for the chart.
v <- c(7,12,28,3,41)
t <- c(14,7,6,19,3)

# Give the chart file a name.
#png(file = "line_chart_2_lines.jpg")

# Plot the bar chart.
plot(v,type = "o",col = "red", xlab = "Month", ylab = "Rain fall", 
   main = "Rain fall chart")

lines(t, type = "o", col = "blue")

# Save the file.
dev.off()

png 
  3

Plot with title “Rain fall chart”

# 散点图
plot(x, y, main, xlab, ylab, xlim, ylim, axes)
以下是所使用的参数的描述 - 
* x是其值为水平坐标的数据集。
* y是其值是垂直坐标的数据集。
* main要是图形的图块。
* xlab是水平轴上的标签。
* ylab是垂直轴上的标签。
* xlim是用于绘图的x的值的极限。
* ylim是用于绘图的y的值的极限。
* axes指示是否应在绘图上绘制两个轴。

In [35]:

# Get the input values.
input <- mtcars[,c('wt','mpg')]

# Give the chart file a name.
#png(file = "scatterplot.png")

# Plot the chart for cars with weight between 2.5 to 5 and mileage between 15 and 30.
plot(x = input$wt,y = input$mpg,
   xlab = "Weight",
   ylab = "Milage",
   xlim = c(2.5,5),
   ylim = c(15,30),		 
   main = "Weight vs Milage"
)
	 
# Save the file.
dev.off()

png 
  3

Plot with title “Weight vs Milage”

# 散点图矩阵
pairs(formula, data)
* formula表示成对使用的一系列变量。
* data表示将从其获取变量的数据集。

In [36]:
# Give the chart file a name.
#png(file = "scatterplot_matrices.png")

# Plot the matrices between 4 variables giving 12 plots.

# One variable with 3 others and total 4 variables.

pairs(~wt+mpg+disp+cyl,data = mtcars,
   main = "Scatterplot Matrix")

# Save the file.
dev.off()

png 
  3

plot without title

# 饼状图
pie(x, labels, radius, main, col, clockwise)
* x是包含饼图中使用的数值的向量。
* labels用于给出切片的描述。
* radius表示饼图圆的半径（值-1和+1之间）。
* main表示图表的标题。
* col表示调色板。
* clockwise是指示片段是顺时针还是逆时针绘制的逻辑值。

In [37]:
# Create data for the graph.
x <- c(21, 62, 10, 53)
labels <- c("London", "New York", "Singapore", "Mumbai")

# Give the chart file a name.
#png(file = "city.jpg")

# Plot the chart.
pie(x,labels)

# Save the file.
dev.off()

png 
  3

plot without title

In [38]:
# Create data for the graph.
x <- c(21, 62, 10, 53)
labels <- c("London", "New York", "Singapore", "Mumbai")

# Give the chart file a name.
#png(file = "city_title_colours.jpg")

# Plot the chart with title and rainbow color pallet.
pie(x, labels, main = "City pie chart", col = rainbow(length(x)))

# Save the file.
dev.off()

png 
  3

Plot with title “City pie chart”

In [39]:
# Create data for the graph.
x <-  c(21, 62, 10,53)
labels <-  c("London","New York","Singapore","Mumbai")

piepercent<- round(100*x/sum(x), 1)

# Give the chart file a name.
#png(file = "city_percentage_legends.jpg")

# Plot the chart.
pie(x, labels = piepercent, main = "City pie chart",col = rainbow(length(x)))
legend("topright", c("London","New York","Singapore","Mumbai"), cex = 0.8,
   fill = rainbow(length(x)))

# Save the file.
dev.off()

png 
  3

Plot with title “City pie chart”

In [47]:
# Get the library.
install.packages("plotrix")
library(plotrix)

# Create data for the graph.
x <-  c(21, 62, 10,53)
lbl <-  c("London","New York","Singapore","Mumbai")

# Give the chart file a name.
#png(file = "3d_pie_chart.jpg")

# Plot the chart.
pie3D(x,labels = lbl,explode = 0.1, main = "Pie Chart of Countries ")

# Save the file.
dev.off()

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


png 
  3

Plot with title “Pie Chart of Countries ”

#  CSV文件
获取和设置工作目录
您可以使用getwd()函数检查R语言工作区指向的目录。 您还可以使用setwd()函数设置新的工作目录。

In [2]:
# Get and print current working directory.
print(getwd())

# Set current working directory.
setwd("/home/kesci/work")

# Get and print current working directory.
print(getwd())

[1] "/home/kesci/work"
[1] "/home/kesci/work"


In [3]:
data <- read.csv("input.csv")
#读取
print(data)

  id     name salary start_date       dept
1  1     Rick 623.30 2012-01-01         IT
2  2      Dan 515.20 2013-09-23 Operations
3  3 Michelle 611.00 2014-11-15         IT
4  4     Ryan 729.00 2014-05-11         HR
5 NA     Gary 843.25 2015-03-27    Finance
6  6     Nina 578.00 2013-05-21         IT
7  7    Simon 632.80 2013-07-30 Operations
8  8     Guru 722.50 2014-06-17    Finance


In [4]:
data <- read.csv("input.csv")
#分析数据结构
print(is.data.frame(data))
print(ncol(data))
print(nrow(data))

[1] TRUE
[1] 5
[1] 8


In [5]:
# Get the max salary from data frame.
sal <- max(data$salary)
print(sal)

[1] 843.25


In [6]:
# Get the person detail having max salary.
retval <- subset(data, salary == max(salary))
print(retval)

  id name salary start_date    dept
5 NA Gary 843.25 2015-03-27 Finance


In [7]:
retval <- subset( data, dept == "IT")
print(retval)

  id     name salary start_date dept
1  1     Rick  623.3 2012-01-01   IT
3  3 Michelle  611.0 2014-11-15   IT
6  6     Nina  578.0 2013-05-21   IT


In [8]:
retval <- subset(data, as.Date(start_date) > as.Date("2014-01-01"))
print(retval)

  id     name salary start_date    dept
3  3 Michelle 611.00 2014-11-15      IT
4  4     Ryan 729.00 2014-05-11      HR
5 NA     Gary 843.25 2015-03-27 Finance
8  8     Guru 722.50 2014-06-17 Finance


In [9]:
#写入CSV文件
# Write filtered data into a new file.
write.csv(retval,"output.csv")
newdata <- read.csv("output.csv")
print(newdata)

  X id     name salary start_date    dept
1 3  3 Michelle 611.00 2014-11-15      IT
2 4  4     Ryan 729.00 2014-05-11      HR
3 5 NA     Gary 843.25 2015-03-27 Finance
4 8  8     Guru 722.50 2014-06-17 Finance


In [10]:
#删除X列
# Create a data frame.
# Write filtered data into a new file.
write.csv(retval,"output.csv", row.names = FALSE)
newdata <- read.csv("output.csv")
print(newdata)

  id     name salary start_date    dept
1  3 Michelle 611.00 2014-11-15      IT
2  4     Ryan 729.00 2014-05-11      HR
3 NA     Gary 843.25 2015-03-27 Finance
4  8     Guru 722.50 2014-06-17 Finance


# Excel文件

In [13]:
install.packages("xlsx")
#不支持java

also installing the dependencies ‘rJava’, ‘xlsxjars’

Warning message in install.packages("xlsx"):
“installation of package ‘rJava’ had non-zero exit status”Warning message in install.packages("xlsx"):
“installation of package ‘xlsxjars’ had non-zero exit status”Warning message in install.packages("xlsx"):
“installation of package ‘xlsx’ had non-zero exit status”Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [14]:
# Verify the package is installed.
any(grepl("xlsx",installed.packages()))

# Load the library into R workspace.
library("xlsx")

[1] FALSE

ERROR: Error in library("xlsx"): there is no package called ‘xlsx’


# R语言 二进制文件
函数WriteBin()和readBin()来创建和读取二进制文件
writeBin(object, con)
readBin(con, what, n )
* con是读取或写入二进制文件的连接对象。
* object是要写入的二进制文件。
* what是模式，如字符，整数等表示要读取的字节。
* n是从二进制文件读取的字节数。

In [19]:
# Read the "mtcars" data frame as a csv file and store only the columns "cyl", "am" and "gear".
write.table(mtcars, file = "mtcars.csv",row.names = FALSE, na = "", 
   col.names = TRUE, sep = ",")

# Store 5 records from the csv file as a new data frame.
new.mtcars <- read.table("mtcars.csv",sep = ",",header = TRUE,nrows = 5)
# Create a connection object to write the binary file using mode "wb".
write.filename = file("/home/kesci/work/binmtcars.dat", "wb")
# Write the column names of the data frame to the connection object.
writeBin(colnames(new.mtcars), write.filename)
# Write the records in each of the column to the file.
writeBin(c(new.mtcars$cyl,new.mtcars$am,new.mtcars$gear), write.filename)
# Close the file for writing so that it can be read by other program.
close(write.filename)

In [21]:
# Create a connection object to read the file in binary mode using "rb".
read.filename <- file("/home/kesci/work/binmtcars.dat", "rb")
# First read the column names. n = 3 as we have 3 columns.
column.names <- readBin(read.filename, character(),  n = 3)
# Next read the column values. n = 18 as we have 3 column names and 15 values.
read.filename <- file("/home/kesci/work/binmtcars.dat", "rb")
bindata <- readBin(read.filename, integer(),  n = 18)
# Print the data.
print(bindata)
# Read the values from 4th byte to 8th byte which represents "cyl".
cyldata = bindata[4:8]
print(cyldata)
# Read the values form 9th byte to 13th byte which represents "am".
amdata = bindata[9:13]
print(amdata)
# Read the values form 9th byte to 13th byte which represents "gear".
geardata = bindata[14:18]
print(geardata)
# Combine all the read values to a dat frame.
finaldata = cbind(cyldata, amdata, geardata)
colnames(finaldata) = column.names
print(finaldata)

 [1]    6778989    7108963 1886611812    7366656 1952543332    7632640
 [7] 1667593073    7566848 1728081249    7496037 1651663203       1536
[13]       1536       1024       1536       2048        256        256
[1]    7366656 1952543332    7632640 1667593073    7566848
[1] 1728081249    7496037 1651663203       1536       1536
[1] 1024 1536 2048  256  256
            mpg        cyl disp
[1,]    7366656 1728081249 1024
[2,] 1952543332    7496037 1536
[3,]    7632640 1651663203 2048
[4,] 1667593073       1536  256
[5,]    7566848       1536  256


#  XML文件

In [22]:
install.packages("XML")

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [23]:
# Load the package required to read XML files.
library("XML")

# Also load the other required package.
library("methods")

# Give the input file name to the function.
result <- xmlParse(file = "input.xml")

# Print the result.
print(result)

<?xml version="1.0"?>
<RECORDS>
  <EMPLOYEE>
    <ID>1</ID>
    <NAME>Rick</NAME>
    <SALARY>623.3</SALARY>
    <STARTDATE>1/1/2012</STARTDATE>
    <DEPT>IT</DEPT>
  </EMPLOYEE>
  <EMPLOYEE>
    <ID>2</ID>
    <NAME>Dan</NAME>
    <SALARY>515.2</SALARY>
    <STARTDATE>9/23/2013</STARTDATE>
    <DEPT>Operations</DEPT>
  </EMPLOYEE>
  <EMPLOYEE>
    <ID>3</ID>
    <NAME>Michelle</NAME>
    <SALARY>611</SALARY>
    <STARTDATE>11/15/2014</STARTDATE>
    <DEPT>IT</DEPT>
  </EMPLOYEE>
  <EMPLOYEE>
    <ID>4</ID>
    <NAME>Ryan</NAME>
    <SALARY>729</SALARY>
    <STARTDATE>5/11/2014</STARTDATE>
    <DEPT>HR</DEPT>
  </EMPLOYEE>
  <EMPLOYEE>
    <ID>5</ID>
    <NAME>Gary</NAME>
    <SALARY>843.25</SALARY>
    <STARTDATE>3/27/2015</STARTDATE>
    <DEPT>Finance</DEPT>
  </EMPLOYEE>
  <EMPLOYEE>
    <ID>6</ID>
    <NAME>Nina</NAME>
    <SALARY>578</SALARY>
    <STARTDATE>5/21/2013</STARTDATE>
    <DEPT>IT</DEPT>
  </EMPLOYEE>
  <EMPLOYEE>
    <ID>7</ID>
    <NAME>Simon</NAME>
    <SALARY>632.8<

In [24]:
#获取XML文件中存在的节点数
# Load the packages required to read XML files.
library("XML")
library("methods")

# Give the input file name to the function.
result <- xmlParse(file = "input.xml")

# Exract the root node form the xml file.
rootnode <- xmlRoot(result)

# Find number of nodes in the root.
rootsize <- xmlSize(rootnode)

# Print the result.
print(rootsize)

[1] 8


In [25]:
#第一个节点的详细信息
# Exract the root node form the xml file.
rootnode <- xmlRoot(result)

# Print the result.
print(rootnode[1])

$EMPLOYEE
<EMPLOYEE>
  <ID>1</ID>
  <NAME>Rick</NAME>
  <SALARY>623.3</SALARY>
  <STARTDATE>1/1/2012</STARTDATE>
  <DEPT>IT</DEPT>
</EMPLOYEE> 

attr(,"class")
[1] "XMLInternalNodeList" "XMLNodeList"        


In [26]:
# Get the first element of the first node.
print(rootnode[[1]][[1]])

# Get the fifth element of the first node.
print(rootnode[[1]][[5]])

# Get the second element of the third node.
print(rootnode[[3]][[2]])

<ID>1</ID> 
<DEPT>IT</DEPT> 
<NAME>Michelle</NAME> 


In [27]:
#将xml文件中的数据作为数据框读取。 
#然后处理数据帧以进行数据分析。
# Convert the input xml file to a data frame.
xmldataframe <- xmlToDataFrame("input.xml")
print(xmldataframe)

  ID     NAME SALARY  STARTDATE       DEPT
1  1     Rick  623.3   1/1/2012         IT
2  2      Dan  515.2  9/23/2013 Operations
3  3 Michelle    611 11/15/2014         IT
4  4     Ryan    729  5/11/2014         HR
5  5     Gary 843.25  3/27/2015    Finance
6  6     Nina    578  5/21/2013         IT
7  7    Simon  632.8  7/30/2013 Operations
8  8     Guru  722.5  6/17/2014    Finance


#  JSON文件

In [28]:
install.packages("rjson")


Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [29]:
# Load the package required to read JSON files.
library("rjson")

# Give the input file name to the function.
result <- fromJSON(file = "input.json")

# Print the result.
print(result)

$ID
[1] "1" "2" "3" "4" "5" "6" "7" "8"

$Name
[1] "Rick"     "Dan"      "Michelle" "Ryan"     "Gary"     "Nina"     "Simon"   
[8] "Guru"    

$Salary
[1] "623.3"  "515.2"  "611"    "729"    "843.25" "578"    "632.8"  "722.5" 

$StartDate
[1] "1/1/2012"   "9/23/2013"  "11/15/2014" "5/11/2014"  "3/27/2015" 
[6] "5/21/2013"  "7/30/2013"  "6/17/2014" 

$Dept
[1] "IT"         "Operations" "IT"         "HR"         "Finance"   
[6] "IT"         "Operations" "Finance"   



In [30]:
# Convert JSON file to a data frame.
json_data_frame <- as.data.frame(result)

print(json_data_frame)

  ID     Name Salary  StartDate       Dept
1  1     Rick  623.3   1/1/2012         IT
2  2      Dan  515.2  9/23/2013 Operations
3  3 Michelle    611 11/15/2014         IT
4  4     Ryan    729  5/11/2014         HR
5  5     Gary 843.25  3/27/2015    Finance
6  6     Nina    578  5/21/2013         IT
7  7    Simon  632.8  7/30/2013 Operations
8  8     Guru  722.5  6/17/2014    Finance


# Web数据

In [31]:
install.packages("RCurl")
install.packages("XML")
install.packages("stringr")
install.packages("plyr")

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [41]:
# Read the URL.
url <- "http://www.geos.ed.ac.uk/~weather/jcmb_ws/"

# Gather the html links present in the webpage.
links <- getHTMLLinks(url)

# Identify only the links which point to the JCMB 2015 files. 
filenames <- links[str_detect(links, "JCMB_2015.csv")]

# Store the file names as a list.
filenames_list <- as.list(filenames)

# Create a function to download the files by passing the URL and filename list.
downloadcsv <- function (mainurl,filename) {
   filedetails <- str_c(mainurl,filename)
   download.file(filedetails,filename)
}

# Now apply the l_ply function and save the files into the current R working directory.
l_ply(filenames,downloadcsv,mainurl = "http://www.geos.ed.ac.uk/~weather/jcmb_ws/")


ERROR: Error: failed to load external entity "http://www.geos.ed.ac.uk/~weather/jcmb_ws/"



# 数据库

In [42]:
install.packages("RMySQL")


Warning message in install.packages("RMySQL"):
“installation of package ‘RMySQL’ had non-zero exit status”Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [46]:
# Create a connection Object to MySQL database.
# We will connect to the sampel database named "sakila" that comes with MySql installation.
mysqlconnection = dbConnection(MySQL(), user = 'root', password = '', dbname = 'sakila',
   host = 'localhost')

# List the tables available in this database.
 dbListTables(mysqlconnection)

ERROR: Error in dbConnection(MySQL(), user = "root", password = "", dbname = "sakila", : could not find function "dbConnection"


In [47]:
# Query the "actor" tables to get all the rows.
result = dbSendQuery(mysqlconnection, "select * from actor")

# Store the result in a R data frame object. n = 5 is used to fetch first 5 rows.
data.frame = fetch(result, n = 5)
print(data.frame)

ERROR: Error in dbSendQuery(mysqlconnection, "select * from actor"): could not find function "dbSendQuery"


In [48]:
result = dbSendQuery(mysqlconnection, "select * from actor where last_name = 'TORN'")

# Fetch all the records(with n = -1) and store it as a data frame.
data.frame = fetch(result, n = -1)
print(data)

ERROR: Error in dbSendQuery(mysqlconnection, "select * from actor where last_name = 'TORN'"): could not find function "dbSendQuery"


In [49]:
#更新查询传递给dbSendQuery()函数来更新Mysql表中的行
dbSendQuery(mysqlconnection, "update mtcars set disp = 168.5 where hp = 110")


ERROR: Error in dbSendQuery(mysqlconnection, "update mtcars set disp = 168.5 where hp = 110"): could not find function "dbSendQuery"


In [50]:
# Use the R data frame "mtcars" to create the table in MySql.
# All the rows of mtcars are taken inot MySql.
dbWriteTable(mysqlconnection, "mtcars", mtcars[, ], overwrite = TRUE)

ERROR: Error in dbWriteTable(mysqlconnection, "mtcars", mtcars[, ], overwrite = TRUE): could not find function "dbWriteTable"


In [51]:
dbSendQuery(mysqlconnection, 'drop table if exists mtcars')

ERROR: Error in dbSendQuery(mysqlconnection, "drop table if exists mtcars"): could not find function "dbSendQuery"


# 平均值
mean(x, trim = 0, na.rm = FALSE, ...)
* x是输入向量。

* trim用于从排序向量的两端丢弃一些观察结果。

* na.rm用于从输入向量中删除缺失值。

In [2]:
# Create a vector. 
x <- c(12,7,3,4.2,18,2,54,-21,8,-5)
# Find Mean.
result.mean <- mean(x)
print(result.mean)

[1] 8.22


当提供trim参数时，向量中的值被排序，然后从计算平均值中减去所需的观察值。

当trim = 0.3时，来自每端的3个值将从计算中减去以找到均值。

In [3]:
# Create a vector.
x <- c(12,7,3,4.2,18,2,54,-21,8,-5)
# Find Mean.
result.mean <-  mean(x,trim = 0.3)
print(result.mean)

[1] 5.55


要从计算中删除缺少的值，请使用na.rm = TRUE。 这意味着去除NA值。

In [4]:
# Create a vector. 
x <- c(12,7,3,4.2,18,2,54,-21,8,-5,NA)

# Find mean.
result.mean <-  mean(x)
print(result.mean)

# Find mean dropping NA values.
result.mean <-  mean(x,na.rm = TRUE)
print(result.mean)

[1] NA
[1] 8.22


# Median中位数
median(x, na.rm = FALSE)
* x是输入向量。
* na.rm用于从输入向量中删除缺失值。

In [5]:
# Create the vector.
x <- c(12,7,3,4.2,18,2,54,-21,8,-5)

# Find the median.
median.result <- median(x)
print(median.result)

[1] 5.6


# Mode
出现次数最多的值

In [6]:
# Create the function.
getmode <- function(v) {
   uniqv <- unique(v)
   uniqv[which.max(tabulate(match(v, uniqv)))]
}

# Create the vector with numbers.
v <- c(2,1,2,3,1,2,3,4,1,5,5,3,2,3)

# Calculate the mode using the user function.
result <- getmode(v)
print(result)

# Create the vector with characters.
charv <- c("o","it","the","it","it")

# Calculate the mode using the user function.
result <- getmode(charv)
print(result)

[1] 2
[1] "it"


# 线性回归
1. 进行收集高度和相应重量的观测值的样本的实验。
2. 使用R语言中的lm()函数创建关系模型。
3. 从创建的模型中找到系数，并使用这些创建数学方程
4. 获得关系模型的摘要以了解预测中的平均误差。 也称为残差。
5. 为了预测新人的体重，使用R中的predict()函数。

lm()函数
lm(formula,data)
* formula是表示x和y之间的关系的符号。
* data是应用公式的向量。

In [1]:
x <- c(151, 174, 138, 186, 128, 136, 179, 163, 152, 131)
y <- c(63, 81, 56, 91, 47, 57, 76, 72, 62, 48)
# Apply the lm() function.
relation <- lm(y~x)
print(relation)


Call:
lm(formula = y ~ x)

Coefficients:
(Intercept)            x  
   -38.4551       0.6746  



In [2]:
x <- c(151, 174, 138, 186, 128, 136, 179, 163, 152, 131)
y <- c(63, 81, 56, 91, 47, 57, 76, 72, 62, 48)

# Apply the lm() function.
relation <- lm(y~x)
#摘要
print(summary(relation))


Call:
lm(formula = y ~ x)

Residuals:
    Min      1Q  Median      3Q     Max 
-6.3002 -1.6629  0.0412  1.8944  3.9775 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) -38.45509    8.04901  -4.778  0.00139 ** 
x             0.67461    0.05191  12.997 1.16e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.253 on 8 degrees of freedom
Multiple R-squared:  0.9548,	Adjusted R-squared:  0.9491 
F-statistic: 168.9 on 1 and 8 DF,  p-value: 1.164e-06



# predict()函数
predict(object, newdata)
* object是已使用lm()函数创建的公式。
* newdata是包含预测变量的新值的向量。

In [3]:
# The predictor vector.
x <- c(151, 174, 138, 186, 128, 136, 179, 163, 152, 131)

# The resposne vector.
y <- c(63, 81, 56, 91, 47, 57, 76, 72, 62, 48)

# Apply the lm() function.
relation <- lm(y~x)

# Find weight of a person with height 170.
a <- data.frame(x = 170)
result <-  predict(relation,a)
print(result)

       1 
76.22869 


In [4]:
# Create the predictor and response variable.
x <- c(151, 174, 138, 186, 128, 136, 179, 163, 152, 131)
y <- c(63, 81, 56, 91, 47, 57, 76, 72, 62, 48)
relation <- lm(y~x)


# Plot the chart.
plot(y,x,col = "blue",main = "Height & Weight Regression",
abline(lm(x~y)),cex = 1.3,pch = 16,xlab = "Weight in Kg",ylab = "Height in cm")

# Save the file.
dev.off()

null device 
          1

Plot with title “Height & Weight Regression”

# 多重回归
y = a + b1x1 + b2x2 +...bnxn
lm(y ~ x1+x2+x3...,data)


In [5]:
#数据集“mtcars”。 它给出了每加仑里程（mpg），气缸排量（“disp”），马力（“hp”），汽车重量（“wt”）
#目标是建立“mpg”作为响应变量与“disp”，“hp”和“wt”作为预测变量之间的关系
input <- mtcars[,c("mpg","disp","hp","wt")]
print(head(input))

                   mpg disp  hp    wt
Mazda RX4         21.0  160 110 2.620
Mazda RX4 Wag     21.0  160 110 2.875
Datsun 710        22.8  108  93 2.320
Hornet 4 Drive    21.4  258 110 3.215
Hornet Sportabout 18.7  360 175 3.440
Valiant           18.1  225 105 3.460


In [6]:
# Create the relationship model.
model <- lm(mpg~disp+hp+wt, data = input)
# Show the model.
print(model)
# Get the Intercept and coefficients as vector elements.
cat("# # # # The Coefficient Values # # # ","")
a <- coef(model)[1]
print(a)
Xdisp <- coef(model)[2]
Xhp <- coef(model)[3]
Xwt <- coef(model)[4]
print(Xdisp)
print(Xhp)
print(Xwt)


Call:
lm(formula = mpg ~ disp + hp + wt, data = input)

Coefficients:
(Intercept)         disp           hp           wt  
  37.105505    -0.000937    -0.031157    -3.800891  

# # # # The Coefficient Values # # #  (Intercept) 
   37.10551 
         disp 
-0.0009370091 
         hp 
-0.03115655 
       wt 
-3.800891 


得出方程:
Y = a+Xdisp.x1+Xhp.x2+Xwt.x3
or
Y = 37.15+(-0.000937)*x1+(-0.0311)*x2+(-3.8008)*x3

# 逻辑回归
y = 1/(1+e^-(a+b1x1+b2x2+b3x3+...))
glm(formula,data,family)
* formula是表示变量之间的关系的符号。
* data是给出这些变量的值的数据集。
* family是R语言对象来指定模型的细节。 它的值是二项逻辑回归。

In [7]:
input <- mtcars[,c("am","cyl","hp","wt")]

am.data = glm(formula = am ~ cyl + hp + wt, data = input, family = binomial)

print(summary(am.data))


Call:
glm(formula = am ~ cyl + hp + wt, family = binomial, data = input)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.17272  -0.14907  -0.01464   0.14116   1.27641  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)  
(Intercept) 19.70288    8.11637   2.428   0.0152 *
cyl          0.48760    1.07162   0.455   0.6491  
hp           0.03259    0.01886   1.728   0.0840 .
wt          -9.14947    4.15332  -2.203   0.0276 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 43.2297  on 31  degrees of freedom
Residual deviance:  9.8415  on 28  degrees of freedom
AIC: 17.841

Number of Fisher Scoring iterations: 8



# 标准分布
dnorm(x, mean, sd)
pnorm(x, mean, sd)
qnorm(p, mean, sd)
rnorm(n, mean, sd)
* x是数字的向量。
* p是概率的向量。
* n是观察的数量（样本大小）。
* mean是样本数据的平均值。 它的默认值为零。
* sd是标准偏差。 它的默认值为1。

In [8]:
#dnorm（）该函数给出给定平均值和标准偏差在每个点的概率分布的高度。
# Create a sequence of numbers between -10 and 10 incrementing by 0.1.
x <- seq(-10, 10, by = .1)

# Choose the mean as 2.5 and standard deviation as 0.5.
y <- dnorm(x, mean = 2.5, sd = 0.5)

plot(x,y)

# Save the file.
dev.off()

null device 
          1

plot without title

In [10]:
#pnorm（）该函数给出正态分布随机数的概率小于给定数的值。 它也被称为“累积分布函数”。
# Create a sequence of numbers between -10 and 10 incrementing by 0.2.
x <- seq(-10,10,by = .2)
 
# Choose the mean as 2.5 and standard deviation as 2. 
y <- pnorm(x, mean = 2.5, sd = 2)

# Plot the graph.
plot(x,y)

# Save the file.
dev.off()

null device 
          1

plot without title

In [11]:
#qnorm（）该函数采用概率值，并给出累积值与概率值匹配的数字。
# Create a sequence of probability values incrementing by 0.02.
x <- seq(0, 1, by = 0.02)

# Choose the mean as 2 and standard deviation as 3.
y <- qnorm(x, mean = 2, sd = 1)

# Plot the graph.
plot(x,y)

# Save the file.
dev.off()

null device 
          1

plot without title

In [13]:
#RNORM（）生成分布正常的随机数。 它将样本大小作为输入，并生成许多随机数。
# Create a sample of 50 numbers which are normally distributed.
y <- rnorm(50)

# Plot the histogram for this sample.
hist(y, main = "Normal DIstribution")

# Save the file.
dev.off()

null device 
          1

Plot with title “Normal DIstribution”

# 二项分布
dbinom(x, size, prob)
pbinom(x, size, prob)
qbinom(p, size, prob)
rbinom(n, size, prob)
* x是数字的向量。
* p是概率向量。
* n是观察的数量。
* size是试验的数量。
* prob是每个试验成功的概率。

In [14]:
#dbinom（）该函数给出每个点的概率密度分布。
# Create a sample of 50 numbers which are incremented by 1.
x <- seq(0,50,by = 1)
# Create the binomial distribution.
y <- dbinom(x,50,0.5)
# Plot the graph for this sample.
plot(x,y)
# Save the file.
dev.off()

null device 
          1

plot without title

In [15]:
#pbinom（）此函数给出事件的累积概率。 它是表示概率的单个值。
# Probability of getting 26 or less heads from a 51 tosses of a coin.
x <- pbinom(26,51,0.5)
print(x)

[1] 0.610116


In [16]:
#qbinom（）该函数采用概率值，并给出累积值与概率值匹配的数字。
# How many heads will have a probability of 0.25 will come out when a coin is tossed 51 times.
x <- qbinom(0.25,51,1/2)
print(x)

[1] 23


In [17]:
#rbinom（）该函数从给定样本产生给定概率的所需数量的随机值。
# Find 8 random values from a sample of 150 with probability of 0.4.
x <- rbinom(8,150,.4)
print(x)

[1] 60 57 63 65 58 59 55 57


# 泊松回归
log(y) = a + b1x1 + b2x2 + bnxn.....
* y是响应变量。
* a和b是数字系数。
* x是预测变量。
glm(formula,data,family)
* formula是表示变量之间的关系的符号。
* data是给出这些变量的值的数据集。
* family是R语言对象来指定模型的细节。 它的值是“泊松”的逻辑回归。

In [1]:
#数据集“warpbreaks”，其描述了羊毛类型（A或B）和张力（低，中或高）
#对每个织机的经纱断裂数量的影响。 让我们考虑“休息”作为响应变量，它是休息次数的计数。
#羊毛“类型”和“张力”作为预测变量。
input <- warpbreaks
print(head(input))

  breaks wool tension
1     26    A       L
2     30    A       L
3     54    A       L
4     25    A       L
5     70    A       L
6     52    A       L


In [2]:
output <-glm(formula = breaks ~ wool+tension, 
                   data = warpbreaks, 
                 family = poisson)
print(summary(output))


Call:
glm(formula = breaks ~ wool + tension, family = poisson, data = warpbreaks)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-3.6871  -1.6503  -0.4269   1.1902   4.2616  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  3.69196    0.04541  81.302  < 2e-16 ***
woolB       -0.20599    0.05157  -3.994 6.49e-05 ***
tensionM    -0.32132    0.06027  -5.332 9.73e-08 ***
tensionH    -0.51849    0.06396  -8.107 5.21e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 297.37  on 53  degrees of freedom
Residual deviance: 210.39  on 50  degrees of freedom
AIC: 493.06

Number of Fisher Scoring iterations: 4



# 协方差分析
我们使用回归分析创建模型，描述变量在预测变量对响应变量的影响。 有时，如果我们有一个类别变量，如Yes / No或Male / Female等。简单的回归分析为分类变量的每个值提供多个结果。 在这种情况下，我们可以通过将分类变量与预测变量一起使用并比较分类变量的每个级别的回归线来研究分类变量的效果。 这样的分析被称为协方差分析，也称为ANCOVA。

In [3]:
#从数据集mtcars创建一个包含字段“mpg”，“hp”和“am”的数据框。 
#这里我们取“mpg”作为响应变量，“hp”作为预测变量，“am”作为分类变量。
input <- mtcars[,c("am","mpg","hp")]
print(head(input))

                  am  mpg  hp
Mazda RX4          1 21.0 110
Mazda RX4 Wag      1 21.0 110
Datsun 710         1 22.8  93
Hornet 4 Drive     0 21.4 110
Hornet Sportabout  0 18.7 175
Valiant            0 18.1 105


In [4]:
#创建一个回归模型，以“hp”作为预测变量，“mpg”作为响应变量，
#考虑“am”和“hp”之间的相互作用。
# Get the dataset.
input <- mtcars
# Create the regression model.
result <- aov(mpg~hp*am,data = input)
print(summary(result))

            Df Sum Sq Mean Sq F value   Pr(>F)    
hp           1  678.4   678.4  77.391 1.50e-09 ***
am           1  202.2   202.2  23.072 4.75e-05 ***
hp:am        1    0.0     0.0   0.001    0.981    
Residuals   28  245.4     8.8                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [5]:
#这个结果表明，马力和传输类型对每加仑的英里有显着的影响，因为两种情况下的p值都小于0.05。 
#但是这两个变量之间的相互作用不显着，因为p值大于0.05。
# Get the dataset.
input <- mtcars
# Create the regression model.
result <- aov(mpg~hp+am,data = input)
print(summary(result))


            Df Sum Sq Mean Sq F value   Pr(>F)    
hp           1  678.4   678.4   80.15 7.63e-10 ***
am           1  202.2   202.2   23.89 3.46e-05 ***
Residuals   29  245.4     8.5                     
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


In [6]:
# Get the dataset.
input <- mtcars

# Create the regression models.
result1 <- aov(mpg~hp*am,data = input)
result2 <- aov(mpg~hp+am,data = input)

# Compare the two models.
print(anova(result1,result2))

Analysis of Variance Table

Model 1: mpg ~ hp * am
Model 2: mpg ~ hp + am
  Res.Df    RSS Df  Sum of Sq     F Pr(>F)
1     28 245.43                           
2     29 245.44 -1 -0.0052515 6e-04 0.9806


# 时间序列分析
时间序列是一系列数据点，其中每个数据点与时间戳相关联。 
一个简单的例子是股票在某一天的不同时间点的股票价格。 
timeseries.object.name <-  ts(data, start, end, frequency)
* data是包含在时间序列中使用的值的向量或矩阵。
* start以时间序列指定第一次观察的开始时间。
* end指定时间序列中最后一次观测的结束时间。
* frequency指定每单位时间的观测数。

In [7]:
# Get the data points in form of a R vector.
rainfall <- c(799,1174.8,865.1,1334.6,635.4,918.5,685.5,998.6,784.2,985,882.8,1071)
# Convert it to a time series object.
rainfall.timeseries <- ts(rainfall,start = c(2012,1),frequency = 12)
# Print the timeseries data.
print(rainfall.timeseries)
# Plot a graph of the time series.
plot(rainfall.timeseries)
# Save the file.
dev.off()

        Jan    Feb    Mar    Apr    May    Jun    Jul    Aug    Sep    Oct
2012  799.0 1174.8  865.1 1334.6  635.4  918.5  685.5  998.6  784.2  985.0
        Nov    Dec
2012  882.8 1071.0


null device 
          1

plot without title

* frequency= 12指定一年中每个月的数据点。

* frequency= 4每年的每个季度的数据点。

* frequency= 6每小时的10分钟的数据点。

* frequency= 24 * 6将一天的每10分钟的数据点固定。

In [8]:
# Get the data points in form of a R vector.
rainfall1 <- c(799,1174.8,865.1,1334.6,635.4,918.5,685.5,998.6,784.2,985,882.8,1071)
rainfall2 <- 
           c(655,1306.9,1323.4,1172.2,562.2,824,822.4,1265.5,799.6,1105.6,1106.7,1337.8)

# Convert them to a matrix.
combined.rainfall <-  matrix(c(rainfall1,rainfall2),nrow = 12)

# Convert it to a time series object.
rainfall.timeseries <- ts(combined.rainfall,start = c(2012,1),frequency = 12)

# Print the timeseries data.
print(rainfall.timeseries)

# Plot a graph of the time series.
plot(rainfall.timeseries, main = "Multiple Time Series")

# Save the file.
dev.off()

         Series 1 Series 2
Jan 2012    799.0    655.0
Feb 2012   1174.8   1306.9
Mar 2012    865.1   1323.4
Apr 2012   1334.6   1172.2
May 2012    635.4    562.2
Jun 2012    918.5    824.0
Jul 2012    685.5    822.4
Aug 2012    998.6   1265.5
Sep 2012    784.2    799.6
Oct 2012    985.0   1105.6
Nov 2012    882.8   1106.7
Dec 2012   1071.0   1337.8


null device 
          1

plot without title

# 非线性最小二乘
大多数时候，真实世界数据模型的方程涉及更高程度的数学函数，如3的指数或sin函数。
线性和非线性回归的目的是调整模型参数的值，以找到最接近您的数据的线或曲线。
在最小二乘回归中，建立了一个回归模型，其中来自回归曲线的不同点的垂直距离的平方和被最小化。
nls(formula, data, start)
* formula是包括变量和参数的非线性模型公式。
* data是用于计算公式中变量的数据框。
* start是起始估计的命名列表或命名数字向量。

In [1]:
xvalues <- c(1.6,2.1,2,2.23,3.71,3.25,3.4,3.86,1.19,2.21)
yvalues <- c(5.19,7.43,6.94,8.11,18.75,14.88,16.06,19.12,3.21,7.58)
# Plot these values.
plot(xvalues,yvalues)
# Take the assumed values and fit into the model.
model <- nls(yvalues ~ b1*xvalues^2+b2,start = list(b1 = 1,b2 = 3))
# Plot the chart with new data by fitting it to a prediction from 100 data points.
new.data <- data.frame(xvalues = seq(min(xvalues),max(xvalues),len = 100))
lines(new.data$xvalues,predict(model,newdata = new.data))
# Save the file.
dev.off()
# Get the sum of the squared residuals.
print(sum(resid(model)^2))
# Get the confidence intervals on the chosen values of the coefficients.
print(confint(model))

null device 
          1

[1] 1.081935


Waiting for profiling to be done...


       2.5%    97.5%
b1 1.137708 1.253135
b2 1.497364 2.496484


plot without title

In [2]:
install.packages("party")


also installing the dependencies ‘mvtnorm’, ‘TH.data’, ‘libcoin’, ‘matrixStats’, ‘multcomp’, ‘modeltools’, ‘strucchange’, ‘coin’, ‘sandwich’

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


决策树是以树的形式表示选择及其结果的图。图中的节点表示事件或选择，并且图的边缘表示决策规则或条件。
它主要用于使用R的机器学习和数据挖掘应用程序。
ctree(formula, data) 
* formula是描述预测变量和响应变量的公式。
* data是所使用的数据集的名称。

In [3]:
#我们将使用名为readingSkills的R内置数据集来创建决策树。 
#它描述了某人的readingSkills的分数，如果我们知道变量“年龄”，“shoesize”，“分数”，以及该人是否为母语者。

# Load the party package. It will automatically load other dependent packages.
library(party)

# Print some records from data set readingSkills.
print(head(readingSkills))

Loading required package: grid
Loading required package: mvtnorm
Loading required package: modeltools
Loading required package: stats4
Loading required package: strucchange
Loading required package: zoo

Attaching package: ‘zoo’

The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric

Loading required package: sandwich


  nativeSpeaker age shoeSize    score
1           yes   5 24.83189 32.29385
2           yes   6 25.95238 36.63105
3            no  11 30.42170 49.60593
4           yes   7 28.66450 40.28456
5           yes  11 31.88207 55.46085
6           yes  10 30.07843 52.83124


In [4]:
# Load the party package. It will automatically load other dependent packages.
library(party)

# Create the input data frame.
input.dat <- readingSkills[c(1:105),]

# Create the tree.
  output.tree <- ctree(
  nativeSpeaker ~ age + shoeSize + score, 
  data = input.dat)

# Plot the tree.
plot(output.tree)

# Save the file.
dev.off()

null device 
          1

plot without title

# 随机森林算法
对构建树时未使用的情况进行错误估计。 这称为OOB（袋外）误差估计，其被提及为百分比。

randomForest(formula, data) 
* formula是描述预测变量和响应变量的公式。
* data是所使用的数据集的名称。

In [3]:
install.packages("randomForest")


Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [4]:
#我们将使用名为readingSkills的R语言内置数据集来创建决策树。 
#它描述了某人的readingSkills的分数，如果我们知道变量“age”，“shoesize”，“score”，以及该人是否是母语。
# Load the party package. It will automatically load other required packages.
library(party)

# Print some records from data set readingSkills.
print(head(readingSkills))


  nativeSpeaker age shoeSize    score
1           yes   5 24.83189 32.29385
2           yes   6 25.95238 36.63105
3            no  11 30.42170 49.60593
4           yes   7 28.66450 40.28456
5           yes  11 31.88207 55.46085
6           yes  10 30.07843 52.83124


In [5]:
# Load the party package. It will automatically load other required packages.
library(party)
library(randomForest)

# Create the forest.
output.forest <- randomForest(nativeSpeaker ~ age + shoeSize + score, 
           data = readingSkills)

# View the forest results.
print(output.forest) 

# Importance of each predictor.
print(importance(fit,type = 2)) 


Call:
 randomForest(formula = nativeSpeaker ~ age + shoeSize + score,      data = readingSkills) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 1

        OOB estimate of  error rate: 1.5%
Confusion matrix:
    no yes class.error
no  99   1        0.01
yes  2  98        0.02


ERROR: Error in UseMethod("importance"): no applicable method for 'importance' applied to an object of class "c('standardGeneric', 'genericFunction', 'function', 'OptionalFunction', 'PossibleMethod', 'functionOrNULL', 'optionalMethod')"


# 生存分析
预测特定事件将要发生的时间。 它也被称为故障时间分析或分析死亡时间。
Surv(time,event)
survfit(formula)

* time是直到事件发生的跟踪时间。
* event指示预期事件的发生的状态。
* formula是预测变量之间的关系。

In [11]:
install.packages("survival")

Updating HTML index of packages in '.Library'
Making 'packages.html' ... done


In [1]:
# Load the library.
library("survival")
# Print first few rows.
print(head(pbc))

# Create the survival object. 
survfit(Surv(pbc$time,pbc$status == 2)~1)


# Plot the graph. 
plot(survfit(Surv(pbc$time,pbc$status == 2)~1))

# Save the file.
dev.off()

  id time status trt      age sex ascites hepato spiders edema bili chol
1  1  400      2   1 58.76523   f       1      1       1   1.0 14.5  261
2  2 4500      0   1 56.44627   f       0      1       1   0.0  1.1  302
3  3 1012      2   1 70.07255   m       0      0       0   0.5  1.4  176
4  4 1925      2   1 54.74059   f       0      1       1   0.5  1.8  244
5  5 1504      1   2 38.10541   f       0      1       1   0.0  3.4  279
6  6 2503      2   2 66.25873   f       0      1       0   0.0  0.8  248
  albumin copper alk.phos    ast trig platelet protime stage
1    2.60    156   1718.0 137.95  172      190    12.2     4
2    4.14     54   7394.8 113.52   88      221    10.6     3
3    3.48    210    516.0  96.10   55      151    12.0     4
4    2.54     64   6121.8  60.63   92      183    10.3     4
5    3.53    143    671.0 113.15   72      136    10.9     3
6    3.98     50    944.0  93.00   63       NA    11.0     3


Call: survfit(formula = Surv(pbc$time, pbc$status == 2) ~ 1)

      n  events  median 0.95LCL 0.95UCL 
    418     161    3395    3090    3853 

null device 
          1

plot without title

# 卡方检验
卡方检验是一种确定两个分类变量之间是否存在显着相关性的统计方法。 
这两个变量应该来自相同的人口，他们应该是类似 - 是/否，男/女，红/绿等。
chisq.test(data)



In [6]:
library("MASS")
print(str(Cars93))

'data.frame':	93 obs. of  27 variables:
 $ Manufacturer      : Factor w/ 32 levels "Acura","Audi",..: 1 1 2 2 3 4 4 4 4 5 ...
 $ Model             : Factor w/ 93 levels "100","190E","240",..: 49 56 9 1 6 24 54 74 73 35 ...
 $ Type              : Factor w/ 6 levels "Compact","Large",..: 4 3 1 3 3 3 2 2 3 2 ...
 $ Min.Price         : num  12.9 29.2 25.9 30.8 23.7 14.2 19.9 22.6 26.3 33 ...
 $ Price             : num  15.9 33.9 29.1 37.7 30 15.7 20.8 23.7 26.3 34.7 ...
 $ Max.Price         : num  18.8 38.7 32.3 44.6 36.2 17.3 21.7 24.9 26.3 36.3 ...
 $ MPG.city          : int  25 18 20 19 22 22 19 16 19 16 ...
 $ MPG.highway       : int  31 25 26 26 30 31 28 25 27 25 ...
 $ AirBags           : Factor w/ 3 levels "Driver & Passenger",..: 3 1 2 1 2 2 2 2 2 2 ...
 $ DriveTrain        : Factor w/ 3 levels "4WD","Front",..: 2 2 2 2 3 2 2 3 2 2 ...
 $ Cylinders         : Factor w/ 6 levels "3","4","5","6",..: 2 4 4 4 2 2 4 4 4 5 ...
 $ EngineSize        : num  1.8 3.2 2.8 2.8 3.5 2.2 3.8 5.7 3.

In [7]:
# Create a data frame from the main data set.
car.data <- data.frame(Cars93$AirBags, Cars93$Type)

# Create a table with the needed variables.
car.data = table(Cars93$AirBags, Cars93$Type) 
print(car.data)

# Perform the Chi-Square test.
print(chisq.test(car.data))

                    
                     Compact Large Midsize Small Sporty Van
  Driver & Passenger       2     4       7     0      3   0
  Driver only              9     7      11     5      8   3
  None                     5     0       4    16      3   6


Warning message in chisq.test(car.data):
“Chi-squared approximation may be incorrect”


	Pearson's Chi-squared test

data:  car.data
X-squared = 33.001, df = 10, p-value = 0.0002723

